keel-core/src/main/kotlin/com/netflix/spinnaker/keel/serialization/ObjectMappers.kt

```diff
@@ -34,14 +42,35 @@ fun configuredYamlMapper(): YAMLMapper =
 
 private fun <T : ObjectMapper> T.configureMe(): T =
   apply {
-    registerKotlinModule()
+    registerModule(DefaultJsonTypeInfoModule)
+      .registerKotlinModule()
       .registerULIDModule()
       .registerModule(JavaTimeModule())
       .configureSaneDateTimeRepresentation()
       .disable(FAIL_ON_UNKNOWN_PROPERTIES)
       .enable(ACCEPT_CASE_INSENSITIVE_ENUMS)
   }
 
+object DefaultJsonTypeInfoModule : SimpleModule() {
+  override fun setupModule(context: SetupContext) {
+    context.insertAnnotationIntrospector(object : NopAnnotationIntrospector() {
+      override fun findTypeResolver(config: MapperConfig<*>, ac: AnnotatedClass, baseType: JavaType): TypeResolverBuilder<*>? {
+        // This is the equivalent of using a @JsonTypeInfo annotation with the specified settings.
+        // We don't want to transitively ship jackson-annotations, though. Sub-types need to be
+        // registered programmatically.
+        return if (baseType.rawClass == Constraint::class.java) {
+          StdTypeResolverBuilder()
+            .init(JsonTypeInfo.Id.NAME, null)
+            .inclusion(JsonTypeInfo.As.EXISTING_PROPERTY)
+            .typeProperty("type")
+        } else {
+          super.findTypeResolver(config, ac, baseType)
+        }
+      }
+    })
+  }
+}
```

Comment:
This is the key bit of Jackson tomfuckery. It's basically replicating what the `@JsonTypeInfo` annotation was doing on the `Constraint` class. It has to be wrapped up in a module so as not to interfere with the Kotlin module (you can just `setAnnotationIntrospector` on `ObjectMapper` but it breaks more things than it fixes).

Reply:
I'm not super-keen on directly referencing `Constraint::class.java` here. I wonder if we should eventually come up with our own annotation or something.